### DISCUSSION TITLE
Updating a subproject within an existing Git repo

### DISCUSSION THREAD

**Discussion by flippedAben** [copier-org/copier#449] [Ideas]
I have a Git repo at `~/repo`. I have another Git repo at `~/my_copier_template`. I tell `copier` to use `~/my_copier_template` to generate a subproject at `~/repo/path/to/S` (call it __S__ for short). I track my new subproject in `~/repo` by adding and committing the changes.

When I try to update __S__, it fails at: https://github.com/copier-org/copier/blob/b8ff4d8c711cf16683dd14f4bf3768df2b85e07e/copier/main.py#L607-L610

This is because `copier` assumes that a subproject is tracked by Git if and only if it is in the root of a git repo, as stated here:
https://github.com/copier-org/copier/blob/b8ff4d8c711cf16683dd14f4bf3768df2b85e07e/copier/subproject.py#L77-L80

But, __S__ is not at the root of `~/repo`. I would like to change copier to support this scenario, but wanted to get some thoughts before proceeding.

Part of the change will involve changing how we update:
https://github.com/copier-org/copier/blob/b8ff4d8c711cf16683dd14f4bf3768df2b85e07e/copier/main.py#L670-L673

How do we feel about not doing the `fetch`, and doing a `git diff` between __R__:`/path/to/S` and the cleanly generated __T__, instead of `git diff-tree`.

**Comment by bburgin** [comment]
+1 for this request.

We want to use copier on a repo with multiple answer files in subfolders. Also, we want to use copier on a large monorepo, but the fetch that copier does for this case is too slow.

**Reply by yajo** [reply]
This is a known problem, but I just noticed I never opened the proper issue for that. Done now: https://github.com/copier-org/copier/issues/450

**Comment by yajo** [comment]
Hi there, all you need to do is put your `copier.yaml` in the git root and add `_subdirectory: repo/path/to/S`.

Docs: https://copier.readthedocs.io/en/latest/configuring/#subdirectory (beware: docs for master, this feature changed from latest v5.x)

**Reply by flippedAben** [reply]
Ah, hold on. There seems to be a misunderstanding. I apologize for the poor wording.

__R__ is a Git repo, but it does not hold `copier` templates. It holds Python source code only, let's say. __R__ is where my generated subproject lives. __T__ is a different Git repo that contains only my copier templates. Within __T__, I already have a `copier.yaml` that has `_subdirectory: path/to/S`. The problem is that when I want to update my generated subproject, `copier` doesn't think my subproject is tracked by a Git repo. But, it is tracked because I committed it to __R__.

I'll edit my initial post to be more clear.

**Reply by yajo** [reply]
Ah, I get your point now.

Ok this seems a valid usage, but indeed not supported yet. I opened https://github.com/copier-org/copier/issues/451 to track it.

PR wellone! 😁

**Comment by bburgin** [comment]
I put a PR up here https://github.com/copier-org/copier/pull/452

**Reply by bburgin** [reply]
Here's a different PR https://github.com/copier-org/copier/pull/453 that better aligns with the copier design and usage.
